### PR TITLE
CodeGen Fix causal mask for half precision

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do? 

This PR forces the causal mask to stay in `torch.uint8`. An error occurs when loading a model in half precision since `torch_dtype=torch.float16` casts also the buffers in fp16. Here is a minimal script to reproduce the error:
```
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("Salesforce/codegen-2B-mono")
model = AutoModelForCausalLM.from_pretrained("Salesforce/codegen-2B-mono", device_map="auto", torch_dtype=torch.float16)

text = "def quicksort(l):"

encoded_input = tokenizer(text, return_tensors='pt')
output_sequences = model.generate(input_ids=encoded_input['input_ids'], attention_mask=encoded_input['attention_mask'])
print(tokenizer.decode(output_sequences[0], skip_special_tokens=True))
```

In a future PR we could address non-casting the buffers (aka keeping them in their native `dtype`)

Can also confirm the slow tests pass!

cc @ydshieh

### CODE REVIEW

File: src/transformers/models/codegen/modeling_codegen.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def _attn(
 
         # compute causal mask from causal mask buffer
         query_length, key_length = query.size(-2), key.size(-2)
-        causal_mask = self.causal_mask[:, :, key_length - query_length : key_length, :key_length]
+        causal_mask = self.causal_mask[:, :, key_length - query_length : key_length, :key_length].to(torch.uint8)
```
Comment (ydshieh) [inline src/transformers/models/codegen/modeling_codegen.py:154]: Let's have a comment here to explain why we need `.to(torch.uint8)` 🙏

Comment (younesbelkada) [reply src/transformers/models/codegen/modeling_codegen.py:158]: Added more comments on 8b81ac12339692bb3bbddc363ddcb9438eeb3e68 
Let me know if anything is unclear!

**Review by sgugger — APPROVED** [review]
LGTM, thanks for fixing!

File: src/transformers/models/codegen/modeling_codegen.py (hunk 1 above)
Comment (thomasw21) [reply src/transformers/models/codegen/modeling_codegen.py:158]: Stupid question, why is that not
```suggestion
        self.causal_mask = self.causal_mask[:, :, key_length - query_length : key_length, :key_length].to(torch.uint8)
```
feels like something you want to do only once no?

Comment (sgugger) [reply src/transformers/models/codegen/modeling_codegen.py:158]: It won't be needed anywhere as we found the root cause with Younes :-)

**Review by sgugger — COMMENTED** [review]
Let's close this one to focus on the right fix @younesbelkada :-)

**Comment by younesbelkada** [comment]
Yeah let's move the discussion to: https://github.com/huggingface/transformers/pull/18471
